UnitTests/GitUITests/CommandsDialogs/CommitDialog/FormCommitTests.cs

```diff
@@ -0,0 +1,209 @@
+            {
+                Assert.AreEqual(generatedCommitMessage, formCommit.GetTestAccessor().Message.Text);
+            });
+        }
+
+        [TestCase(CommitKind.Fixup)]
+        [TestCase(CommitKind.Squash)]
+        public void DoNotPreserveCommitMessageOnReopenFromSpecialCommit(CommitKind commitKind)
+        {
+            var generatedCommitMessage = Path.GetRandomFileName();
+
+            RunFormCommitTest(
+                formCommit =>
+                {
+                    string prefix = commitKind.ToString().ToLowerInvariant();
+                    Assert.AreEqual($"{prefix}! A commit message", formCommit.GetTestAccessor().Message.Text);
+                    formCommit.GetTestAccessor().Message.Text = generatedCommitMessage;
+                },
+                commitKind);
+
+            RunFormCommitTest(formCommit =>
+            {
+                Assert.IsEmpty(formCommit.GetTestAccessor().Message.Text);
+            });
+        }
+
+        [Test]
+        public void SelectMessageFromHistory()
+        {
+            var generatedCommitMessage = Path.GetRandomFileName();
+
+            RunFormCommitTest(formCommit =>
+            {
+                var commitMessageToolStripMenuItem = formCommit.GetTestAccessor().CommitMessageToolStripMenuItem;
+
+                // Verify the message appears correctly
+                commitMessageToolStripMenuItem.ShowDropDown();
+                Assert.AreEqual("A commit message...", commitMessageToolStripMenuItem.DropDownItems[0].Text);
+
+                // Verify the message is selected correctly
+                commitMessageToolStripMenuItem.DropDownItems[0].PerformClick();
+                Assert.AreEqual("A commit message", formCommit.GetTestAccessor().Message.Text);
+            });
+        }
+
+        private static async Task WaitForIdleAsync()
+        {
+            var idleCompletionSource = new TaskCompletionSource<VoidResult>();
+            Application.Idle += HandleApplicationIdle;
+
+            // Queue an event to make sure we don't stall if the application was already idle
+            await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
+            await Task.Yield();
+
+            await idleCompletionSource.Task;
+            Application.Idle -= HandleApplicationIdle;
+
+            void HandleApplicationIdle(object sender, EventArgs e)
+            {
+                idleCompletionSource.TrySetResult(default);
+            }
+        }
+
+        private void RunFormCommitTest(Action<FormCommit> testDriver, CommitKind commitKind = CommitKind.Normal)
+        {
+            RunFormCommitTest(
+                formCommit =>
+                {
+                    testDriver(formCommit);
+                    return Task.CompletedTask;
+                },
+                commitKind);
+        }
+
+        private void RunFormCommitTest(Func<FormCommit, Task> testDriverAsync, CommitKind commitKind = CommitKind.Normal)
```

Comment:
💭 It may be possible to further extract a shared test helper `RunFormTest`, but it didn't seem necessary for this initial implementation.